pkg/usagestats/seed_test.go

```diff
@@ -0,0 +1,317 @@
+	newSeed := ClusterSeed{UID: "new", CreatedAt: now}
+
+	tests := map[string]struct {
+		setup               func(bucketClient objstore.Bucket)
+		expectedErr         error
+		expectedMinDuration time.Duration
+	}{
+		"should immediately return if seed file exists and it was created more than 'min stability' time ago": {
+			setup: func(bucketClient objstore.Bucket) {
+				data, err := json.Marshal(oldSeed)
+				require.NoError(t, err)
+				require.NoError(t, bucketClient.Upload(context.Background(), ClusterSeedFileName, bytes.NewReader(data)))
+			},
+			expectedMinDuration: 0,
+		},
+		"should wait for 'min stability' and return the seed file if it exists and was created less than 'min stability' time ago": {
+			setup: func(bucketClient objstore.Bucket) {
+				data, err := json.Marshal(newSeed)
+				require.NoError(t, err)
+				require.NoError(t, bucketClient.Upload(context.Background(), ClusterSeedFileName, bytes.NewReader(data)))
+			},
+			expectedMinDuration: minStability,
+		},
+		"should create the seed file if doesn't exist and then wait for 'min stability'": {
+			setup:               func(bucketClient objstore.Bucket) {},
+			expectedMinDuration: minStability,
+		},
+		"should re-create the seed file if exist but is corrupted, and then wait for 'min stability'": {
+			setup: func(bucketClient objstore.Bucket) {
+				require.NoError(t, bucketClient.Upload(context.Background(), ClusterSeedFileName, bytes.NewReader([]byte("xxx"))))
+			},
+			expectedMinDuration: minStability,
+		},
+	}
+
+	for testName, testData := range tests {
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			bucketClient, err := filesystem.NewBucketClient(filesystem.Config{Directory: t.TempDir()})
+			require.NoError(t, err)
+
+			testData.setup(bucketClient)
+
+			startTime := time.Now()
+			actualSeed, err := initSeedFile(context.Background(), objstore.BucketWithMetrics("", bucketClient, nil), minStability, log.NewNopLogger())
+			if testData.expectedErr != nil {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), testData.expectedErr.Error())
+			} else {
+				require.NoError(t, err)
+
+				// We expect the seed stored in the bucket.
+				expectedSeed, err := readSeedFile(context.Background(), objstore.BucketWithMetrics("", bucketClient, nil), log.NewNopLogger())
+				require.NoError(t, err)
+				require.Equal(t, expectedSeed.UID, actualSeed.UID)
+				require.Equal(t, expectedSeed.CreatedAt.Unix(), actualSeed.CreatedAt.Unix())
+			}
+
+			require.GreaterOrEqual(t, time.Since(startTime), testData.expectedMinDuration)
+		})
+	}
+}
+
+func TestInitSeedFile_CreatingConcurrency(t *testing.T) {
+	t.Parallel()
+
+	const (
+		numReplicas  = 100
+		minStability = 5 * time.Second
+	)
+
+	var (
+		start   = make(chan struct{})
+		seedsMx sync.Mutex
+		seeds   = make([]ClusterSeed, 0, numReplicas)
+	)
+
+	bucketClient, err := filesystem.NewBucketClient(filesystem.Config{Directory: t.TempDir()})
+	require.NoError(t, err)
```

Comment:
I think this test would be _better_ if we wrapped the bucket client with something that adds a:

```go
defer func() {
        time.Sleep(
                time.Duration(rand.Int63n(int64(5 * time.Millisecond))) +
                time.Millisecond
        )
}() 
```

in each call.

This way we'd exaggerate the real world scenario where the status of the bucket may have changed already when you're getting the response.